### PR TITLE
[python3] Add sitecustomize on windows for dll lookup

### PR DESCRIPTION
**Pull request by Neumann-A**

### CODE REVIEW

**Review by FrankXie05 — DISMISSED** [review]

**Comment by BillyONeal** [comment]
Can you explain why you are doing this? Tools generally need to be copied to tools with their needed DLLs, not reach outside of tools.

**Comment by Neumann-A** [comment]
> Can you explain why you are doing this? Tools generally need to be copied to tools with their needed DLLs, not reach outside of tools.

Ownership problems. If I build numpy -> numpy requires lapack but some other python extension might also require lapack directly. As such numpy cannot copy lapack dlls because it will create a collision. 
Also this avoids needing to copy all of /bin to tools/python3/DLLs (or tools/python3) 

Of course people who do final deployment can copy whatever they want.

**Comment by BillyONeal** [comment]
The problem is that anything that causes loader ownership behavior to change is a huge security mess as it opens 'file lie in wait' holes. It also creates problems because the python 'in vcpkg' is run in a way that the end user would never ever do since nobody deploys their app and expects their dependencies in "my_app/../../bin".

I agree that it sucks that we don't have a great way to model this pattern right now but patching tools to dig for DLLs in other places is... scary. This is fewer patches than in https://github.com/microsoft/vcpkg/pull/31279 but the impacts are pretty similar.

**Comment by Neumann-A** [comment]
@BillyONeal: I mean I can add a check that `sys.executable` is actually python from vcpkg? (Meaning that it is located in a directory called `tools/python3`)

**Comment by Neumann-A** [comment]
My point is: Just tell me what checks to implement and I'll do it. I mean this is a lot better than having to patch python since it allows for the user to simply delete the file before deployment and it uses one of the mechanism python itself implements.

File: ports/python3/portfile.cmake [hunk 1]
```diff
@@ -379,6 +379,17 @@ endif()
...
 if(VCPKG_TARGET_IS_WINDOWS)
   vcpkg_replace_string("${CURRENT_PACKAGES_DIR}/tools/python3/Lib/distutils/command/build_ext.py" "'libs'" "'../../lib'")
+      file(WRITE "${CURRENT_PACKAGES_DIR}/tools/python3/Lib/site-packages/sitecustomize.py"
+[[
+import os
+import sys
+from pathlib import Path
+
+vcpkg_bin_path = Path(sys.prefix + '/../../bin')
+if vcpkg_bin_path.is_dir():
```
Comment (Neumann-A) [inline ports/python3/portfile.cmake:389]:
```suggestion
if sys.executable.endswith('\\tools\\python3\\python.exe') and Path(sys.prefix + '/../../bin').is_dir():
```
@BillyONeal: Would this be better?

Comment (Neumann-A) [reply ports/python3/portfile.cmake:389]: there could also be an `else` block which prints a message to remove the file on deployment if it gets accidently deployed.

**Comment by JavierMatosD** [comment]
> @BillyONeal: I mean I can add a check that `sys.executable` is actually python from vcpkg? (Meaning that it is located in a directory called `tools/python3`)

This wouldn't eliminate the risk of malicious DLLs being placed in the extra directories that are searched

**Comment by Neumann-A** [comment]
> This wouldn't eliminate the risk of malicious DLLs being placed in the extra directories that are searched. 

The point is that the extra directories should be within the vcpkg installed tree. If we don't trust the vcpkg installed tree then even speaking about security issues due to this PR seems strange to me.

**Comment by JavierMatosD** [comment]
> > This wouldn't eliminate the risk of malicious DLLs being placed in the extra directories that are searched.
> 
> The point is that the extra directories should be within the vcpkg installed tree. If we don't trust the vcpkg installed tree then even speaking about security issues due to this PR seems strange to me.

What does this enable for you? For example, is this in order to be able to `cd vcpkg_installed/tools/python/python.exe` ... `import numpy`? If so, how is that necessary to build cpp libraries?

**Comment by Neumann-A** [comment]
https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindPython/Support.cmake?ref_type=heads#L3796-L3801 otherwise you would be required to pass `_${_PYTHON_PREFIX}_NumPy_INCLUDE_DIR` or `_${_PYTHON_PREFIX}_NumPy_PATH` manually. 
https://gitlab.kitware.com/cmake/cmake/-/blob/master/Modules/FindPython/Support.cmake?ref_type=heads#L3819
looks for the version of numpy. typically meson would do something similar in the lookup without a proper way to pass this stuff manually.

There is also 
https://github.com/microsoft/vcpkg/pull/36270#issuecomment-1935100803
https://github.com/microsoft/vcpkg/pull/36270#issuecomment-1935120504

secondly, making sure that installed python extensions actually work in the end should require at least one import to test it.
The import on linux already works due to RPATH lookup, however there is no equivalent on Windows and PATH can also no longer be used.

File: ports/python3/portfile.cmake [hunk 2]
```diff
@@ -379,6 +379,23 @@ endif()
...
+from pathlib import Path
+
+vcpkg_bin_path = Path(sys.prefix + '/../../bin')
+if sys.executable.endswith('\\tools\\python3\\python.exe') and vcpkg_bin_path.is_dir():
+  os.add_dll_directory(vcpkg_bin_path)
+else:
+  raise RuntimeError(f" \n \
+   ******************************************************\n \
+   {sys.executable} seems to be running outside of vcpkg.\n \
+   {os.path.abspath(__file__)} is not ment to be deployed, please delete it! \n \
```
Comment (m-kuhn) [inline ports/python3/portfile.cmake:395]:
```suggestion
   {os.path.abspath(__file__)} is not meant to be deployed, please delete it! \n \
```

**Comment by m-kuhn** [comment]
Waiting on this as well: `gdal[python]`, `qscintilla[python]`

**Comment by vicroms** [comment]
Relitigating this one tomorrow.

**Comment by BillyONeal** [comment]
@vicroms
@AugP
@data-queue
@ras0219-msft
@BillyONeal
@JavierMatosD

@BillyONeal So my understanding is that we have this tradeoff between enabling Python to run from the installed tree, vs. opening this security hole that every application that ships vcpkg python how has to guard/ACL "../bin" and friends.

@ras0219-msft I am concerned about the potential direction question here. If we don't believe that's a problem we are interested in resolving then resolving technical limitations is unnecessary.

Questions:

* Is it a design goal that the python port in vcpkg can provide a full python distribution which is runnable from the installed directory?
* Do the scenarios intended to be solved by this PR actually need a build script, meaning executable on the host and not the target, to load python modules built by vcpkg?


@JavierMatosD I thought that omniorbpy was a C++ library with Python capabilities.
@ras0219-msft My understanding is that omniorb is a C++ library and the one with py provides python bindings to it.
@JavierMatosD Is that a scenario we are interested in supporting? @BillyONeal I observe we provided Boost.Python et al for years; what changed that forces us to need this hack?
@ras0219-msft This hack is about making "cd PYTHONPATHininstalledtree, python import numpy" work, and we need to understand specifically why that's necessary. I just haven't seen the concrete use case here. @Neumann-A linked some lines but those suggest we could just set the CMake variables.
@JavierMatosD But that might not work from Meson
@ras0219-msft What that would mean is 'it is impossible to cross compile for Python with meson' because in cross compiling scenarios you can't run python from the installed tree, full stop. So if meson is really doing this that seems broken and we need to find some way to pass anything like this. Maybe that's a bug report for Meson.
@ras0219-msft I agree with what @JavierMatosD wrote previously ( https://github.com/microsoft/vcpkg/pull/38297#issuecomment-2078299006 ) and I feel like the questions raised therein have not been answered yet.

>secondly, making sure that installed python extensions actually work in the end should require at least one import to test it. The import on linux already works due to RPATH lookup, however there is no equivalent on Windows and PATH can also no longer be used.

@BillyONeal I think the question is why this scenario is expected to work in the first place, not whether or not your PR is the correct change in order to achieve that @Neumann-A

@ras0219-msft Correct, if we wanted to test imports or something we could do what we tell customers to do and make their own bespoke custom deployment for that test. And we need to know why we can't do what we're telling customers to do in the first place before discussing the merits of one technical way of achieving that or another.

**Comment by Neumann-A** [comment]
> And we need to know why we can't do what we're telling customers to do in the first place before

Because venv changes all the internal paths python normally returns to the one of the venv. Again python on !windows just works as expected. I only need a way to make it work on windows. I can add a function `get_installed_python_env` which manually creates a `venv` without actually requiring `venv` being installed and basically installing a similar sitecustomize as here, creating a `pyvenv.cfg` and copying the python executable and the runtime and dlls into the `BUILDTREE`. 
Want me to pursue that approach @ras0219-msft @BillyONeal? 

> What that would mean is 'it is impossible to cross compile for Python with meson' because in cross compiling scenarios you can't run python from the installed tree, full stop.

cross compiling with python is a whole different story. That requires `crossenv` which only works on `!windows` since the file it is looking for does not exist on windows. using `crossenv` entails taking a host/build and a target python and running the host with the paths of the target python.  

> I observe we provided Boost.Python et al for years; what changed that forces us to need this hack?

Boost.Python does not provide python bindings. It provides C++ bindings to python stuff (which than can be C/C++ stuff). The installation is also incomplete since it is missing the numpy bindings (which of course would require the numpy headers and libs).  

@ras0219-msft @BillyONeal: Furthermore would it be ok to create a port for the changes here so that a person wanting a working  in tree python can actually have it?

**Pull request closed without merging**